### PR TITLE
[Organizations] Functional tests for push/unlist of organization packages

### PR DESCRIPTION
**Pull request by scottbommarito**
This adds
- UI tests
    - Push existing package
    - Push new package as organization admin
    - Push new version of an existing package as organization admin
    - Push new version of an existing package as organization collaborator
- CLI tests
    - Push new package as organization admin
    - Push and unlist new version of an existing package as organization admin
    - Push and unlist new version of an existing package as organization collaborator

No negative test cases (like the rest of our functional tests) but I can add some if we determine we want them.

Note that for this to run we will need to
- onboard the test account to two organizations (one as admin, one as collaborator)
- create two API keys, one scoped to each organization
- create two packages, one owned by each organization
- add the names of the organizations, their API keys, and their package IDs to the configuration of the functional tests

### CODE REVIEW

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/AssertAndValidationHelper.cs [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 namespace NuGetGallery.FunctionalTests.Helpers
 {
-    public class AssertAndValidationHelper
+    public static class AssertAndValidationHelper
```
Comment (scottbommarito) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/AssertAndValidationHelper.cs]: Just for context that may not be immediately obvious...

This file was in the WebUITests.P0 project but was used by all three WebUITests projects through a direct file reference. I moved it to the Core project and then updated the references accordingly.

File: tests/NuGetGallery.FunctionalTests.Core/EnvironmentSettings.cs [hunk 2]
```diff
@@ -27,6 +27,12 @@ public class EnvironmentSettings
         private static string _testAccountApiKey_Unlist;
         private static string _testAccountApiKey_PushPackage;
         private static string _testAccountApiKey_PushVersion;
+        private static string _testOrganizationAdminAccountName;
+        private static string _testOrganizationAdminAccountPackageId;
+        private static string _testOrganizationAdminAccountApiKey;
+        private static string _testOrganizationCollaboratorAccountName;
+        private static string _testOrganizationCollaboratorAccountPackageId;
```
Comment (chenriksson) [inline tests/NuGetGallery.FunctionalTests.Core/EnvironmentSettings.cs]: Can the same package ID be used for org admin and collaborator?

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/UploadHelper.cs [hunk 3]
```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using Microsoft.VisualStudio.TestTools.WebTesting;
+using System.Linq;
+using System;
```
Comment (chenriksson) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/UploadHelper.cs]: nit: sort usings

File: tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs [hunk 4]
```diff
@@ -48,14 +49,49 @@ public NugetCommandLineTests(ITestOutputHelper testOutputHelper)
...
+        [Priority(0)]
+        [Category("P0Tests")]
+        public async Task UploadAndUnlistNewVersionOfPackageAsOrganizationCollaboratorWithNuGetCommandLineTest()
+        {
+            var id = EnvironmentSettings.TestOrganizationCollaboratorAccountPackageId;
+            var version = UploadHelper.GetUniquePackageVersion();
+            var apiKey = EnvironmentSettings.TestOrganizationCollaboratorAccountApiKey;
+            await _clientSdkHelper.UploadNewPackageAndVerify(id, version, apiKey: apiKey);
+            await _clientSdkHelper.UnlistPackageAndVerify(id, version, apiKey);
+        }
```
Comment (chenriksson) [inline tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: Add test to verify that collaborator can't push new registration?

File: tests/NuGetGallery.WebUITests.P0/UploadAndDownload/UploadPackageWithNewIdFromUI.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    public class UploadPackageWithNewIdFromUI : WebTest
+    {
+        public UploadPackageWithNewIdFromUI()
+        {
+            PreAuthenticate = true;
+        }
+
+        public override IEnumerator<WebTestRequest> GetRequestEnumerator()
+        {
+            return UploadHelper.UploadPackage(this, EnvironmentSettings.TestAccountName);
```
Comment (chenriksson) [inline tests/NuGetGallery.WebUITests.P0/UploadAndDownload/UploadPackageWithNewIdFromUI.cs]: Instead of refactoring into a helper, why not just make the existing test configurable?

Not sure what's possible with WebTest, but you could always create an abstract base class, then override with role-based configuration.

```
namespace NuGetGallery.FunctionalTests.WebUITests.UploadAndDownload
{
  public class UploadAsOrganizationAdminFromUI : UploadPackageFromUI
  {
    public override AccountName => EnvironmentSettings.TestOrganizationAdminAccountName;
    public override PackageId => EnvironmentSettings.TestOrganizationAdminAccountPackageId;
  }

  public class UploadAsSelfFromUI : UploadPackageFromUI
  {
    public override AccountName => EnvironmentSettings.TestAccountName;
    public override PackageId => EnvironmentSettings.TestAccountPackageId;
  }

  public abstract class UploadPackageFromUI : WebTest
  {
    public abstract string AccountName { get; }
    public abstract string PackageId { get; }
    public virtual string PackageVersion { get { return "1.0.0"; } }
    
    // mostly existing code
    public override IEnumerator<WebTestRequest> GetRequestEnumerator() { ... }
  }
}
```

File: tests/NuGetGallery.FunctionalTests.Core/EnvironmentSettings.cs (hunk 2 above)
Comment (skofman1) [reply tests/NuGetGallery.FunctionalTests.Core/EnvironmentSettings.cs]: Package Id doesn't need to be configurable. Use a constant/auto generate.

File: tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs [hunk 6]
```diff
@@ -48,14 +49,49 @@ public NugetCommandLineTests(ITestOutputHelper testOutputHelper)
...
         public async Task UploadPackageWithNuGetCommandLineTest()
         {
             await _clientSdkHelper.UploadNewPackageAndVerify(DateTime.Now.Ticks.ToString());
         }
 
+        [Fact]
+        [Description("Creates a test package with a new ID and pushes it to the server using Nuget.exe as an organization admin")]
+        [Priority(0)]
+        [Category("P0Tests")]
+        public async Task UploadAndUnlistNewPackageAsOrganizationAdminWithNuGetCommandLineTest()
```
Comment (skofman1) [inline tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: The test name is "UploadAndUnlist", but it only updates.

File: tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs [hunk 7]
```diff
@@ -48,14 +49,49 @@ public NugetCommandLineTests(ITestOutputHelper testOutputHelper)
...
+        {
+            var id = EnvironmentSettings.TestOrganizationAdminAccountPackageId;
+            var version = UploadHelper.GetUniquePackageVersion();
+            var apiKey = EnvironmentSettings.TestOrganizationAdminAccountApiKey;
+            await _clientSdkHelper.UploadNewPackageAndVerify(id, version, apiKey: apiKey);
+            await _clientSdkHelper.UnlistPackageAndVerify(id, version, apiKey);
+        }
+
+        [Fact]
+        [Description("Creates a new version of an existing test package and pushes it to the server using Nuget.exe as an organization admin")]
```
Comment (skofman1) [inline tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: >admin [](start = 135, length = 5)

Can we have a single test for Admins, and a single test for Collaborator? 
For admin you can: publish new package, publish new version, unlist.
For collaborator, you will verify publish new package fails, publish new version succeeds, unlist succeeds.

Also, those test have to test the negative scenarios, not only the positive ones.

File: tests/NuGetGallery.FunctionalTests.Core/EnvironmentSettings.cs (hunk 2 above)
Comment (scottbommarito) [reply tests/NuGetGallery.FunctionalTests.Core/EnvironmentSettings.cs]: Can't autogenerate. Constant sounds good tho.

File: tests/NuGetGallery.WebUITests.P0/UploadAndDownload/UploadPackageWithNewIdFromUI.cs (hunk 5 above)
Comment (scottbommarito) [reply tests/NuGetGallery.WebUITests.P0/UploadAndDownload/UploadPackageWithNewIdFromUI.cs]: I'm using `GetUniquePackageVersion` in the CLI tests so I need this helper to share the methods between the two.

Comment (scottbommarito) [reply tests/NuGetGallery.WebUITests.P0/UploadAndDownload/UploadPackageWithNewIdFromUI.cs]: restructured the testing into a class structure as suggested however

File: tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs (hunk 6 above)
Comment (scottbommarito) [reply tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: Fixed.

File: tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (scottbommarito) [reply tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: fixed

Comment (scottbommarito) [reply tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: done

File: tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs (hunk 4 above)
Comment (scottbommarito) [reply tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: done

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/UploadHelper.cs (hunk 3 above)
Comment (scottbommarito) [reply tests/NuGetGallery.FunctionalTests.Core/Helpers/UploadHelper.cs]: done

File: tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs [hunk 8]
```diff
@@ -48,14 +49,54 @@ public NugetCommandLineTests(ITestOutputHelper testOutputHelper)
...
+        [Category("P0Tests")]
+        public async Task UploadAndUnlistPackagesAsAsOrganizationAdminWithNuGetCommandLineTest()
+        {
+            var apiKey = EnvironmentSettings.TestOrganizationAdminAccountApiKey;
+
+            // Can push new package ID as organization
+            await _clientSdkHelper.UploadNewPackageAndVerify(DateTime.Now.Ticks.ToString(), apiKey: apiKey);
+
+            // Can push new version of an existing package as organization
+            var id = Constants.TestOrganizationAdminPackageId;
```
Comment (skofman1) [inline tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: you don't actually need the hardcoded id here. You can push to the same registration you created two lines higher.

Comment (scottbommarito) [reply tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: true, my bad

collaborator's package needs to be constant however

Comment (skofman1) [reply tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: You won't need UploaderHelper either.. push version 1.0.0 in the first line, version 2.0.0 here. Unlist either version.

---
In reply to: [164903835](https://github.com/NuGet/NuGetGallery/pull/5362#discussion_r164903835) [](ancestors = 164903835)

File: tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs [hunk 9]
```diff
@@ -48,14 +49,54 @@ public NugetCommandLineTests(ITestOutputHelper testOutputHelper)
...
+        [Category("P0Tests")]
+        public async Task UploadAndUnlistPackagesAsOrganizationCollaboratorWithNuGetCommandLineTest()
+        {
+            var apiKey = EnvironmentSettings.TestOrganizationCollaboratorAccountApiKey;
+
+            // Cannot push new package ID as organization
+            await _clientSdkHelper.UploadNewPackage(DateTime.Now.Ticks.ToString(), success: false, apiKey: apiKey);
+
+            // Can push new version of an existing package as organization
+            var id = Constants.TestOrganizationCollaboratorPackageId;
```
Comment (skofman1) [inline tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: suggestion: your life would be even easier (and you can avoid pre-uploading packages), if you combine the 2 tests into 1. It will do this:
1. Push new package id X 1.0.0 as admin.
2. Push new version of X 2.0.0 as admin
3. Unlist X 1.0.0 as admin
4. Try push Y as collaborator (fail)
5. Push X 3.0.0 as collaborator
6. Unlist X 3.0.0 as collaborator

File: tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs [hunk 10]
```diff
@@ -70,66 +111,47 @@ public NugetCommandLineTests(ITestOutputHelper testOutputHelper)
...
-                Assert.True(processResult.ExitCode != 0, "Package push succeeded, although was expected to fail.");
-
-                // 2. Try to upload package using 'push version' api key => expect failure
-                TestOutputHelper.WriteLine($"2. Trying to upload package '{packageId}', version '{version1}' using 'push version' API key. Expected result: failure.");
-                processResult = await commandlineHelper.UploadPackageAsync(package1FullPath, UrlHelper.V2FeedPushSourceUrl, EnvironmentSettings.TestAccountApiKey_PushVersion);
-                Assert.True(processResult.ExitCode != 0, "Package push succeeded, although was expected to fail.");
+            // 1. Try to upload package using 'unlist' api key => expect failure
+            TestOutputHelper.WriteLine($"1. Trying to upload package '{packageId}', version '{version1}' using 'unlist' API key. Expected result: failure.");
+            await _clientSdkHelper.UploadNewPackage(packageId, version1, success: false, apiKey: EnvironmentSettings.TestAccountApiKey_Unlist);
```
Comment (skofman1) [inline tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: why did u change this code?

File: tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs (hunk 9 above)
Comment (scottbommarito) [reply tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: Unfortunately not true. The same user cannot be both an admin and a collaborator of an organization. Therefore, to avoid creating multiple user accounts, I set up these tests such that there are two organizations that the test user is a member of--one as an admin and one as a collaborator. Therefore I can't push one package as an admin and use that package for collaborator tests because the ownership is different.

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs [hunk 11]
```diff
@@ -150,55 +149,59 @@ public bool CheckIfPackageExistsInSource(string packageId, string sourceUrl)
         /// Creates a package with the specified Id and Version and uploads it and checks if the upload has succeeded.
         /// Throws if the upload fails or cannot be verified in the source.
         /// </summary>
-        public async Task UploadNewPackageAndVerify(string packageId, string version = "1.0.0", string minClientVersion = null, string title = null, string tags = null, string description = null, string licenseUrl = null, string dependencies = null)
+        public async Task UploadNewPackageAndVerify(string packageId, string version = "1.0.0", string minClientVersion = null, string title = null, string tags = null, string description = null, string licenseUrl = null, string dependencies = null, string apiKey = null)
         {
-            await UploadNewPackage(packageId, version, minClientVersion, title, tags, description, licenseUrl, dependencies);
+            await UploadNewPackage(packageId, version, true, minClientVersion, title, tags, description, licenseUrl, dependencies, apiKey);
```
Comment (chenriksson) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: It's not apparent from the caller what the `true` argument is for. Could we return Task<bool> for the status code, or use a named argument?

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs [hunk 12]
```diff
@@ -150,55 +149,59 @@ public bool CheckIfPackageExistsInSource(string packageId, string sourceUrl)
...
 
             await VerifyPackageExistsInV2AndV3Async(packageId, version);
         }
 
-        public async Task UploadNewPackage(string packageId, string version = "1.0.0", string minClientVersion = null,
+        public async Task UploadNewPackage(string packageId, string version = "1.0.0", bool success = true, string minClientVersion = null,
             string title = null, string tags = null, string description = null, string licenseUrl = null,
             string dependencies = null, string apiKey = null)
         {
-            if (string.IsNullOrEmpty(packageId))
```
Comment (chenriksson) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: Why remove this?

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs [hunk 13]
```diff
@@ -150,55 +149,59 @@ public bool CheckIfPackageExistsInSource(string packageId, string sourceUrl)
...
             var packageCreationHelper = new PackageCreationHelper(TestOutputHelper);
             var packageFullPath = await packageCreationHelper.CreatePackage(packageId, version, minClientVersion, title, tags, description, licenseUrl, dependencies);
 
-            await UploadExistingPackage(packageFullPath);
-
-            // Delete package from local disk once it gets uploaded
-            CleanCreatedPackage(packageFullPath);
-        }
-
-        public async Task UploadExistingPackage(string packageFullPath, string apiKey = null)
```
Comment (chenriksson) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: It may still be useful to keep `UploadExistingPackage(string path, string apiKey)` separate.

Could you return `Task<int>` with nuget.exe exit code instead, and check it from `UploadNewPackage`?

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs [hunk 14]
```diff
@@ -150,55 +149,59 @@ public bool CheckIfPackageExistsInSource(string packageId, string sourceUrl)
...
 
         /// <summary>
         /// Unlists a package with the specified Id and Version and checks if the unlist has succeeded.
         /// Throws if the unlist fails or cannot be verified in the source.
         /// </summary>
-        public async Task UnlistPackageAndVerify(string packageId, string version = "1.0.0")
+        public async Task UnlistPackageAndVerify(string packageId, string version = "1.0.0", string apiKey = null)
         {
-            await UnlistPackage(packageId, version);
+            await UnlistPackage(packageId, version, apiKey: apiKey);
```
Comment (chenriksson) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: nit: named arg not necessary since variable name is good

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs [hunk 15]
```diff
@@ -150,55 +149,59 @@ public bool CheckIfPackageExistsInSource(string packageId, string sourceUrl)
...
 
             await VerifyPackageExistsInV2AndV3Async(packageId, version);
         }
 
-        public async Task UnlistPackage(string packageId, string version = "1.0.0", string apiKey = null)
+        public async Task UnlistPackage(string packageId, string version = "1.0.0", bool success = true, string apiKey = null)
```
Comment (chenriksson) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: nit: would list `success` arg last, since we likely want callers to specify it as named arg

File: tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs [hunk 16]
```diff
@@ -48,88 +49,119 @@ public NugetCommandLineTests(ITestOutputHelper testOutputHelper)
         }
 
         [Fact]
-        [Description("Creates a test package and pushes it to the server using Nuget.exe")]
+        [Description("Tests upload and unlist scenarios as self")]
         [Priority(0)]
         [Category("P0Tests")]
-        public async Task UploadPackageWithNuGetCommandLineTest()
+        public async Task UploadAndUnlistPackagesAsSelfWithNuGetCommandLineTest()
```
Comment (chenriksson) [inline tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: nit: `WithNuGetCommandLineTest` seems redundant with class name; don't think it's necessary

File: tests/NuGetGallery.WebUITests.P0/UploadAndDownload/UploadPackageFromUI.cs [hunk 17]
```diff
@@ -1,57 +1,24 @@
...
 using NuGetGallery.FunctionalTests.Helpers;
 
 namespace NuGetGallery.FunctionalTests.WebUITests.UploadAndDownload
 {
-    /// <summary>
-    /// Uploads a new test package using Gallery UI. Validates that logon prompt appears to upload and checks that the package's home page opens post upload.
-    /// priority : p0
-    /// </summary>
-    public class UploadPackageFromUI : WebTest
+    public abstract class UploadPackageFromUI : WebTest
```
Comment (chenriksson) [inline tests/NuGetGallery.WebUITests.P0/UploadAndDownload/UploadPackageFromUI.cs]: These tests look nice!

File: tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs [hunk 18]
```diff
@@ -48,88 +49,119 @@ public NugetCommandLineTests(ITestOutputHelper testOutputHelper)
...
-                await _clientSdkHelper.UnlistPackage(packageId, version2, EnvironmentSettings.TestAccountApiKey_Unlist);
-
-            }
-            finally
-            {
-                _clientSdkHelper.CleanCreatedPackage(package1FullPath);
-                _clientSdkHelper.CleanCreatedPackage(package2FullPath);
-            }
+            TestOutputHelper.WriteLine($"8. Trying to unlist package '{packageId}', version '{version2}' using 'unlist' API key. Expected result: success.");
+            await _clientSdkHelper.UnlistPackage(packageId, version2, apiKey: EnvironmentSettings.TestAccountApiKey_Unlist);
```
Comment (chenriksson) [inline tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: nit: `apiKey` named arg not necessary since it's clear from the environment setting

File: tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs [hunk 19]
```diff
@@ -48,88 +49,119 @@ public NugetCommandLineTests(ITestOutputHelper testOutputHelper)
...
-                Assert.True(processResult.ExitCode != 0, "Package delete succeeded, although was expected to fail.");
+            // 7. Unlist both packages using 'unlist' api key => expect succees
+            TestOutputHelper.WriteLine($"7. Trying to unlist package '{packageId}', version '{version1}' using 'unlist' API key. Expected result: success.");
+            await _clientSdkHelper.UnlistPackage(packageId, version1, apiKey: EnvironmentSettings.TestAccountApiKey_Unlist);
 
-                // 6. Try unlisting package version2 using 'push version' api key => expect failture
-                TestOutputHelper.WriteLine($"6. Trying to unlist package '{packageId}', version '{version2}' using 'push' API key. Expected result: failure.");
-                processResult = await commandlineHelper.DeletePackageAsync(packageId, version2, UrlHelper.V2FeedPushSourceUrl, EnvironmentSettings.TestAccountApiKey_PushVersion);
-                Assert.True(processResult.ExitCode != 0, "Package delete succeeded, although was expected to fail.");
```
Comment (chenriksson) [inline tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: Was `DeletePackageAsync` replaced with one of the `UnlistPackage` calls? Just wondering if a different API is getting called now?

Comment (chenriksson) [reply tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: Could you clarify what changed in `VerifyScopedApiKeysWithNuGetCommandLineTest`, as it doesn't initially seem related to organizations.

File: tests/NuGetGallery.FunctionalTests.Core/EnvironmentSettings.cs [hunk 20]
```diff
@@ -27,6 +27,10 @@ public class EnvironmentSettings
         private static string _testAccountApiKey_Unlist;
         private static string _testAccountApiKey_PushPackage;
         private static string _testAccountApiKey_PushVersion;
+        private static string _testOrganizationAdminAccountName;
+        private static string _testOrganizationAdminAccountApiKey;
+        private static string _testOrganizationCollaboratorAccountName;
+        private static string _testOrganizationCollaboratorAccountApiKey;
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply tests/NuGetGallery.FunctionalTests.Core/EnvironmentSettings.cs]: To clarify: the test account is a member of this organization. This is not a user that is a collaborator of some organization. There is no password because an organization cannot be logged into.

File: tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs (hunk 16 above)
Comment (scottbommarito) [reply tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: It was existing--I will remove.

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs (hunk 12 above)
Comment (scottbommarito) [reply tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: It was unused.

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs (hunk 13 above)
Comment (scottbommarito) [reply tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: I don't see much value keeping them separate--it calls an API that can be called directly instead.

I think it might be worthwhile changing this method to return the `ProcessResult` however.

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs (hunk 11 above)
Comment (scottbommarito) [reply tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: I didn't want to make this a named argument because it would require all the others to be named, but I think I'll just move this to the end of the parameters.

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs [hunk 21]
```diff
@@ -1,14 +1,13 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using NuGet;
+using NuGet.Versioning;
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: need to fix my VS configuration, woops

File: tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs (hunk 19 above)
Comment (scottbommarito) [reply tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: It's an existing test that I refactored to use the improvements in the testing infrastructure I made for the organizations tests.

I could revert it, but I think this code needs a lot of stylistic improvements in general and I tested this one enough to be sure that I didn't change behavior.

**Comment by scottbommarito** [comment]
@chenriksson @skofman1 all PR comments responded to

File: tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs (hunk 19 above)
Comment (chenriksson) [reply tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: I'm fine if it's just refactoring to use the new code... it's just hard to review changes to this test.

Mostly, I wanted to know if there was behavior change. For instance, I see `commandlineHelper.DeletePackageAsync` in the removed code but not in the new code.

Comment (scottbommarito) [reply tests/NuGetGallery.FunctionalTests/Commandline/NuGetCommandLineTests.cs]: I think it might be easier with whitespace diffing off.

Basically, we now call `ClientSdkHelper.UploadPackageAsync` and `ClientSdkHelper.UnlistPackageAsync` which call `CommandHelper.UploadPackageAsync` and `CommandLineHelper.DeletePackageAsync` instead of calling the later directly. The new methods do some setup logic that this test previously did itself.
